components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AbstractApplicationAuthenticator.java

```diff
@@ -179,17 +187,32 @@ protected void publishAuthenticationStepAttempt(HttpServletRequest request, Auth
             if (success) {
                 authnDataPublisherProxy.publishAuthenticationStepSuccess(request, context,
                         unmodifiableParamMap);
+                // Resetting the authenticator start time to null since the step event is Success and for the next
+                // step event start time will be added in DefaultStepHandler handle method.
+                context.setAnalyticsData(FrameworkConstants.AnalyticsData.CURRENT_AUTHENTICATOR_START_TIME, null);
 
             } else {
                 authnDataPublisherProxy.publishAuthenticationStepFailure(request, context,
                         unmodifiableParamMap);
+                // Resetting the authenticator start time to current time since the step event is failure and retrying
+                // the event duration will be counted as a new step.
+                context.setAnalyticsData(
+                        FrameworkConstants.AnalyticsData.CURRENT_AUTHENTICATOR_START_TIME, System.currentTimeMillis());
             }
         }
     }
 
+    private void publishAuthenticationStepAttemptFailure(HttpServletRequest request, AuthenticationContext context,
```

Comment:
It is better to add a doc comment for this method even if it is private.

Reply:
Fixed with #3159